### ISSUE TITLE
Do we still export unused "Custom" axes to designspace file?

### ISSUE THREAD

**Issue by behdad** [googlei18n/glyphsLib#336]
I thought we fixed that. I'm seeing:
```
    <axes> 
        <axis default="400.0" maximum="900.0" minimum="100.0" name="weight" tag="wght"> 
            <map input="100.0" output="26.0" /> 
            <map input="200.0" output="39.0" /> 
            <map input="300.0" output="58.0" /> 
            <map input="400.0" output="90.0" /> 
            <map input="500.0" output="108.0" /> 
            <map input="600.0" output="128.0" /> 
            <map input="700.0" output="151.0" /> 
            <map input="800.0" output="169.0" /> 
            <map input="900.0" output="190.0" /> 
            <labelname xml:lang="en">Weight</labelname> 
        </axis> 
        <axis default="100.0" maximum="100.0" minimum="70.0" name="width" tag="wdth"> 
            <map input="70.0" output="70.0" /> 
            <map input="79.0" output="79.0" /> 
            <map input="89.0" output="89.0" /> 
            <map input="100.0" output="100.0" /> 
            <labelname xml:lang="en">Width</labelname> 
        </axis> 
        <axis default="0.0" maximum="0.0" minimum="0.0" name="custom" tag="XXXX"> 
            <map input="0.0" output="0.0" /> 
            <labelname xml:lang="en">Custom</labelname> 
        </axis> 
    </axes> 
```

I believe I updated my checkout just today. Did I mess up?

**Comment by schriftgestalt** [comment]
The point is what it shouldn’t always create those there axes. Glyphs has a property to define the axes now and if that is not defined, it should check the masters and make axes if there are any differences. In your example, the Custom axis shouldn’t be there.

**Comment by belluzj** [comment]
Sorry, I probably introduced that regression. We're looking into it

**Comment by madig** [comment]
Can you provide a Glyphs file? We can't replicate this here...

**Comment by anthrotype** [comment]
yea, can't reproduce either, using current master, with glyphs file containing Axes custom parameter and not.
I think this if statement that guards the `self.designspace.addAxis(axis)` already does the job:
https://github.com/googlei18n/glyphsLib/blob/5133a43a2c8652633935525f27c98059fdbc71a2/Lib/glyphsLib/builder/axes.py#L178-L184

**Comment by belluzj** [comment]
I think we need to look at the .glyphs file, because maybe there's a difference between the instance mapping and the master mapping for the custom axis, like for example the instance mapping is "empty" in the business sense `len() == 1` but not empty in the python sense `len() == 0` so it is selected to be written out by 
https://github.com/googlei18n/glyphsLib/blob/5133a43a2c8652633935525f27c98059fdbc71a2/Lib/glyphsLib/builder/axes.py#L163-L164
but then the master mapping maybe has `len() > 1` for some reason (e.g. there's a master that has a customValue != 0) and that's why the axis is written out to the file

**Comment by anthrotype** [comment]
I don't think any noto sources uses the third axis, they're all weight + width only. What should I grep for?

**Comment by belluzj** [comment]
I don't know really, I was hoping @behdad could say which file gave that output?

**Comment by anthrotype** [comment]
ok, there's no rush

**Comment by anthrotype** [comment]
This seems to be fixed in glyphsLib master. The "Custom" axis with default == minimum == maximum is not being exported to the generated designspace. 
However, the map element of width in the example above is also redundant, because all the mappings are identical and so there is no "warp" of user space to design space coordinates.

**Comment by madig** [comment]
Should map elements not be written if input and output is the same? Denis mentioned you might need to write them all out if even a single entry differs? The design space document doesn't say when/if they're required...

**Comment by anthrotype** [comment]
> need to write them all out if even a single entry differs?

yes, only when all are equal it can be omitted. The map elements is to construct avar table in varLib.
I'm working on that too, let's not duplicate efforts.

**Comment by anthrotype** [comment]
fixed in #343
